net: sockets: socketpair: fix socketpair coverity issues

cfriedt opened the pull request:
I have no permissions to access the actual errors on coverity.com (error 401), but please let me know if this helps.

Fixes #25793
Fixes #25788
Fixes #25780
Fixes #25737
Fixes #25731 
Fixes #25736 
Fixes #25727 
Fixes #25797

jukkar reviewed (changes requested):
Thanks for the fixes, they look good. We usually add the Coverity CID information to commits that fix these things.
Could you add to each commit this line before/after the "Fixes" line:
```
Coverity-CID: xxxx
```
so that we can find from git log whether a certain Coverity issue was fixed or not.

cfriedt committed:
net: socket: socketpair: remove dead code

In this, case is_nonblock is false and will_block is true.
Therefore, we *may* block, and furthermore we *expect* to
block. Checking is_nonblock is, in fact, redundant, and
passing K_FOREVER to k_sem_take() is justified.

Fixes #25727
Coverity-CID: 210611

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt committed:
tests: socket: socketpair: assert recvfrom returns >= 0

Fixes #25731
Coverity-CID: 210568

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt committed:
samples: sockets: socketpair: check return value of send

Fixes #25736
Coverity-CID: 210583

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt committed:
samples: sockets: socketpair: check return of pthread_create

Fixes #25737
Coverity-CID: 210585

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt committed:
samples: sockets: socketpair: mitigate negative index

Fixes #25780
Coverity-CID: 210612

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt committed:
net: socket: socketpair: initialize after NULL check

Fixes #25788
Coverity-CID: 210581

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt committed:
tests: socket: socketpair: init param before passing by ref

Fixes #25796
Coverity-CID: 210579

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt committed:
tests: socket: socketpair: init param before passing by ref

Fixes #25797
Coverity-CID: 210607

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt commented:
Should be good now. Does Coverity mention when a problem is fixed on GitHub? I'll keep an eye on it in any case.

cfriedt commented:
@jukkar - should be good to go

carlescufi commented:
> @pfalcon, @jukkar, @tbursztyka, @carlescufi - should be good to go

Thanks! waiting for the reviewers above to approve

pfalcon commented on subsys/net/lib/sockets/socketpair.c:
```diff
@@ -476,19 +476,11 @@ static ssize_t spair_write(void *obj, const void *buffer, size_t count)
 			}
 
-			res = k_sem_take(&remote->sem, K_NO_WAIT);
+			res = k_sem_take(&remote->sem, K_FOREVER);
```
It's Coverity in its case 210611 said that K_NO_WAIT here needs to be replaced in K_FOREVER, and called it all "dead code"? That seems want the commit message suggests, or I don't understand it.

pabigot replied:
Yeah, I'm not convinced this is right.  Coverity analyzed a path where this would be reached in a condition where `is_nonblock` is false (because if it wasn't it'd have jumped to `out`).  In that case the "check-no-wait, if fail check other things and if not fail wait forever" reduces to "wait forever".  I don't believe that path is the only path that reaches this point, and in others `is_nonblock` would be true.

I'd try removing that by factoring out the no-wait/check/wait-forever sequence, which appears at least twice in this function, to a separate function.  And I'd do that in a separate PR because it's messy.

cfriedt replied:
It's a bit confusing. I'll have another look soon

cfriedt replied:
@pabigot, @pfalcon - I think Coverity might actually be right in this case.

So the conditions that get is until this loop are 
1. `is_nonblock == false` (i.e. we *may* block)
1. `avail == 0` (i.e. no space to write, so we *will* block, hence `will_block == true`).

So for all intents and purposes, we are legitimately allowed to 
* wait for a signal on `&remote->read_signal` (which blocks)
* attempt to pull up the file object pointed to by the (integer) remote fd (which might be NULL, if it was deleted from under our noses)
* attempt to take a semaphore (which could take forever since we *may* block and in fact we *expect* to block)

If `avail > 0`, `will_block` is false, and a write will occur immediately (regardless of the value of `is_nonblock`), then we do not go into the loop.

However, there is a ticket open for a scenario that I had not initially checked for - which would be reading to or writing from a socket in ISR context (it sounds comical, but someone _might_ actually try to do that and doing so should always fail if the socket is blocking). That is issue #25417 which I'll get to when we're done here.

pfalcon replied:
@cfriedt: My concern, as I mentioned, is to have clear enough and detailed enough commit description. That's how I review stuff - read commit description, and compare with what code changes actually do. Any discrepancies aren't helpful. (And this review may happen now or in a year, when investigating some (un)related issue.)

cfriedt replied:
@pfalcon - commit message has been updated.

stephanosio commented on tests/net/socket/socketpair/src/test_socketpair_happy_path.c:
```diff
@@ -115,10 +115,12 @@ static void happy_path(
 			NULL, &len);
 
 		zassert_not_equal(res, -1, "recvfrom(2) failed: %d", errno);
+		zassert_true(res >= 0, "recvfrom must return >= -1: %d", res);
```
```suggestion
		zassert_true(res >= 0, "recvfrom must return >= 0: %d", res);
```

cfriedt replied:
Technically it can return -1 if there is an error. It should never return -2 etc though

stephanosio replied:
I thought you put `res >= 0` because you have `zassert_not_equal(res, -1, ...)` right above i.t

pabigot commented on tests/net/socket/socketpair/src/test_socketpair_happy_path.c:
```diff
@@ -119,6 +119,7 @@ static void happy_path(
 			NULL, &len);
 
 		zassert_not_equal(res, -1, "recvfrom(2) failed: %d", errno);
```
The original isn't quite redundant with the new check, but do you need both?  Why was `-1` special before?

Also: This can't be 210568, which is "Argument cannot be negative"

cfriedt replied:
I'll have another look again shortly

cfriedt replied:
The assert message is different. In the `!= -1` case, we print `errno`, and in the `>= -1` case, we're checking for valid range of the return. I could add comments to that effect if you think it's more clear.

cfriedt replied:
@pabigot - It is actually 201568. The issue is that the length argument to `strncmp()` cannot be negative because it's implicitly cast to `size_t`. So, the additional assertion is there to specifically prevent that corner case).
If you prefer, I can combine the two assertions to check for `>= 0' and also print `errno`.

cfriedt replied:
The 80-character limit only leaves so much room for messages unfortunately. In Linux 100 is the new limit ;-)

pabigot commented on samples/net/sockets/socketpair/src/socketpair_example.c:
```diff
@@ -47,7 +47,12 @@ static const char *const names[] = {
 {
 	/* write(2) should be used after #25443 */
-	send(fd, name, strlen(name), 0);
+	int res = send(fd, name, strlen(name), 0);
```
This checks the return value as Coverity wants, but it's not a good example IMO because surely the failure should be returned to the caller and checked there.

Not blocking, net maintainers should weigh in.

cfriedt replied:
Since it's sample code, I'm mostly impartial here. I just thought it would might be useful to print an error message to notify the user who would, presumably, be watching.

pabigot commented on samples/net/sockets/socketpair/src/socketpair_example.c:
```diff
@@ -110,8 +111,10 @@ int main(int argc, char *argv[])
-		socketpair(AF_UNIX, SOCK_STREAM, 0, ctx[i].spair);
-		pthread_create(&ctx[i].thread, NULL, fun, &ctx[i]);
+		r = socketpair(AF_UNIX, SOCK_STREAM, 0, ctx[i].spair);
+		assert(r == 0);
```
This should probably be `__ASSERT_NO_MSG()`, which is the Zephyr way of doing this.

cfriedt replied:
Yeah, that won't work for the Makefile.posix way of building though.

pabigot replied:
Why not?  There are uses of Zephyr `__ASSERT()` in lib/posix.

cfriedt replied:
@pabigot, Makefile.posix is for running this sample code entirely without Zephyr as a native Linux (or BSD, or macOS) application. I.e. lib/posix is nonexistent.

cfriedt replied:
A possible alternative solution that doesn't use any asserts is to just use a goto to go to the end of the function. There is no resource cleanup, but again, it's just a sample application to illustrate usage.

pabigot replied:
Or, if you really don't care, use `(void)socketpair(...)`.

cfriedt replied:
I should probably care. Just remembered that `pthread_create()` requires a `pthread_attr_t` in Zephyr, so that needs to be modified anyway.

pfalcon replied:
I think assert() is fine here, and @cfriedt, thanks for providing Makefile.posix - that's the way we can ensure that we actually implement POSIX API portable across OSes. (Maybe one sweet day we'll be able to do the same for tests, but examples with Makefile.posix are the baseline to stick to.)

pfalcon replied:
@cfriedt: Also, FYI (not just for this code piece, but in general), there's an option of marking Coverity case as false positive. I myself don't remember practicing it much, but see e.g. @jukkar practicing it regularly (no implications meant). So, if somewhere your code is correct, but Coverity doesn't get it, you can comment on the initial ticket that it's a false positive (someone would need to propagate that back to Coverity UI).

pabigot commented on samples/net/sockets/socketpair/src/socketpair_example.c:
```diff
@@ -144,6 +144,10 @@ int main(int argc, char *argv[])
 
 			fd = fds[i].fd;
 			idx = fd_to_idx(fd, ctx, ARRAY_SIZE(ctx));
+			if (idx < 0) {
```
This might be 210612.  It's not 210585.

cfriedt replied:
I'll have another look soon.

cfriedt replied:
Absolutely right. Thanks for catching that.

pabigot commented on tests/net/socket/socketpair/src/test_socketpair_unsupported_calls.c:
```diff
@@ -58,6 +58,9 @@ void test_socketpair_unsupported_calls(void)
 			"listen should set errno to EINVAL");
 
 		res = accept(sv[i], (struct sockaddr *)&addr, &len);
```
IMO we don't need to do this; it's test code, and if the accept() didn't fail it's fine to leak the descriptor.

cfriedt replied:
Ok. I wasn't sure if it was mandatory. Will have a look again soon

cfriedt replied:
dropped the commit

pabigot commented on tests/net/socket/socketpair/src/test_socketpair_happy_path.c:
```diff
@@ -115,6 +115,7 @@ static void happy_path(
 
 		memset(actual_msg, 0, sizeof(actual_msg));
 
+		len = 0;
```
A better fix would be to initialize the variable at its declaration up on line 48, as Coverity notes:
```
   	14. var_decl: Declaring variable len without initializer.
 48        socklen_t len;
```

cfriedt replied:
Ah, ok. It is a loop though, so it should be initialized twice (once for each of the two iterations) before passing by reference.

pabigot replied:
No; it's actually not used at all, because `src_addr` is NULL.  But Coverity doesn't know that, so we need to give it a value just once.

cfriedt replied:
Except Coverity doesn't get that 😭 You & I do, of course. Reinitializing `len` before the function is called is technically more correct-ish, should mitigate the Coverity issue, and that's all I really care about.

pabigot reviewed (changes requested):
Some of the associations between changes and Coverity IDs don't make sense.  Other comments need changes too; I think the ones that are optional note that.

pfalcon commented:
> Some of the associations between changes and Coverity IDs don't make sense. Other comments need changes too; I think the ones that are optional note that.

Well, my primary attention is to the changes outside tests/ and samples/. Changes to those can be treated in more or less lax way, but for changes to the system functional code, would be nice to have clear and detailed commit messages.

cfriedt commented:
> > Some of the associations between changes and Coverity IDs don't make sense. Other comments need changes too; I think the ones that are optional note that.
> 
> Well, my primary attention is to the changes outside tests/ and samples/. Changes to those can be treated in more or less lax way, but for changes to the system functional code, would be nice to have clear and detailed commit messages.

@pabigot, @pfalcon - I believe most of your concerns should be addressed. If there is anything else, please let me know.
